Fix cardinality of opentelemetry span names on client-side navigation

PatNeedham opened the pull request:
<!-- Thanks for opening a PR! Your contribution is much appreciated.
To make sure your PR is handled as smoothly as possible we request that you follow the checklist sections below.
Choose the right checklist for the change(s) that you're making:

## For Contributors

This is a duplicate of https://github.com/vercel/next.js/pull/59282 in order to (hopefully) regenerate the logs from the failing tests so that the original PR can be appropriately modified and then merged.

### Improving Documentation

- Run `pnpm prettier-fix` to fix formatting issues before opening the PR.
- Read the Docs Contribution Guide to ensure your contribution follows the docs guidelines: https://nextjs.org/docs/community/contribution-guide

### Adding or Updating Examples

- The "examples guidelines" are followed from our contributing doc https://github.com/vercel/next.js/blob/canary/contributing/examples/adding-examples.md
- Make sure the linting passes by running `pnpm build && pnpm lint`. See https://github.com/vercel/next.js/blob/canary/contributing/repository/linting.md

### Fixing a bug

- Related issues linked using `fixes #number`
- Tests added. See: https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md#writing-tests-for-nextjs
- Errors have a helpful link attached, see https://github.com/vercel/next.js/blob/canary/contributing.md

### Adding a feature

- Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR. (A discussion must be opened, see https://github.com/vercel/next.js/discussions/new?category=ideas)
- Related issues/discussions are linked using `fixes #number`
- e2e tests added (https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md#writing-tests-for-nextjs)
- Documentation added
- Telemetry added. In case of a feature if it's used or not.
- Errors have a helpful link attached, see https://github.com/vercel/next.js/blob/canary/contributing.md


## For Maintainers

- Minimal description (aim for explaining to someone not on the team to understand the PR)
- When linking to a Slack thread, you might want to share details of the conclusion
- Link both the Linear (Fixes NEXT-xxx) and the GitHub issues
- Add review comments if necessary to explain to the reviewer the logic behind a change

### What?

### Why?

### How?

Closes NEXT-
Fixes #

-->

ijjk commented:
Allow CI Workflow Run

- [ ] approve CI run for commit: 878cf30b2050d6ec24bb8969f6bc0bf5f583e176

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

ijjk commented:
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | PatNeedham/next.js fix/telemetry-client-navigation | Change |
| - | - | - | - |
| buildDuration | 18.2s | 16.6s | N/A |
| buildDurationCached | 9.5s | 8.7s | N/A |
| nodeModulesSize | 360 MB | 360 MB | N/A |
| nextStartRea..uration (ms) | 453ms | 446ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | PatNeedham/next.js fix/telemetry-client-navigation | Change |
| - | - | - | - |
| 1103-HASH.js gzip | 31.9 kB | 31.9 kB | N/A |
| 1a9f679d-HASH.js gzip | 53.5 kB | 53.5 kB | N/A |
| 335-HASH.js gzip | 5.09 kB | 5.09 kB | N/A |
| 7953.HASH.js gzip | 169 B | 169 B | ✓ |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 231 B | 228 B | N/A |
| main-HASH.js gzip | 31.5 kB | 31.5 kB | N/A |
| webpack-HASH.js gzip | 1.65 kB | 1.65 kB | N/A |
| Overall change | 45.4 kB | 45.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | PatNeedham/next.js fix/telemetry-client-navigation | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | PatNeedham/next.js fix/telemetry-client-navigation | Change |
| - | - | - | - |
| _app-HASH.js gzip | 192 B | 193 B | N/A |
| _error-HASH.js gzip | 192 B | 193 B | N/A |
| amp-HASH.js gzip | 507 B | 511 B | N/A |
| css-HASH.js gzip | 341 B | 343 B | N/A |
| dynamic-HASH.js gzip | 2.52 kB | 2.52 kB | ✓ |
| edge-ssr-HASH.js gzip | 266 B | 265 B | N/A |
| head-HASH.js gzip | 362 B | 365 B | N/A |
| hooks-HASH.js gzip | 392 B | 392 B | ✓ |
| image-HASH.js gzip | 4.32 kB | 4.32 kB | N/A |
| index-HASH.js gzip | 268 B | 268 B | ✓ |
| link-HASH.js gzip | 2.69 kB | 2.7 kB | N/A |
| routerDirect..HASH.js gzip | 329 B | 328 B | N/A |
| script-HASH.js gzip | 397 B | 397 B | ✓ |
| withRouter-HASH.js gzip | 324 B | 324 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 4 kB | 4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | PatNeedham/next.js fix/telemetry-client-navigation | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | PatNeedham/next.js fix/telemetry-client-navigation | Change |
| - | - | - | - |
| index.html gzip | 528 B | 526 B | N/A |
| link.html gzip | 541 B | 539 B | N/A |
| withRouter.html gzip | 524 B | 521 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | PatNeedham/next.js fix/telemetry-client-navigation | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94.7 kB | 94.7 kB | N/A |
| page.js gzip | 181 kB | 181 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | PatNeedham/next.js fix/telemetry-client-navigation | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 622 B | 626 B | N/A |
| middleware-r..fest.js gzip | 156 B | 156 B | ✓ |
| middleware.js gzip | 25.7 kB | 25.7 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 995 B | 995 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | PatNeedham/next.js fix/telemetry-client-navigation | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | N/A |
| app-page-exp..prod.js gzip | 98.4 kB | 98.4 kB | N/A |
| app-page-tur..prod.js gzip | 99.9 kB | 99.9 kB | N/A |
| app-page-tur..prod.js gzip | 94.3 kB | 94.3 kB | ✓ |
| app-page.run...dev.js gzip | 157 kB | 157 kB | N/A |
| app-page.run..prod.js gzip | 93 kB | 93 kB | N/A |
| app-route-ex...dev.js gzip | 21.5 kB | 21.5 kB | ✓ |
| app-route-ex..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route.ru...dev.js gzip | 21.3 kB | 21.3 kB | ✓ |
| app-route.ru..prod.js gzip | 15 kB | 15 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 21.5 kB | 21.5 kB | ✓ |
| pages.runtim...dev.js gzip | 22.1 kB | 22.1 kB | ✓ |
| pages.runtim..prod.js gzip | 21.5 kB | 21.5 kB | ✓ |
| server.runti..prod.js gzip | 51.6 kB | 51.6 kB | ✓ |
| Overall change | 343 kB | 343 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | PatNeedham/next.js fix/telemetry-client-navigation | Change |
| - | - | - | - |
| 0.pack gzip | 1.62 MB | 1.62 MB | ⚠️ +458 B |
| index.pack gzip | 112 kB | 112 kB | N/A |
| Overall change | 1.62 MB | 1.62 MB | ⚠️ +458 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..ntime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: c9f8807bfaa44c1b56723782379a1fcfe5a581bb

ijjk commented:
## Failing test suites <!--23260044732-->
Commit: c9f8807bfaa44c1b56723782379a1fcfe5a581bb

`pnpm test-start test/production/graceful-shutdown/index.test.ts`
- Graceful Shutdown > production (standalone mode) > should not accept new requests during shutdown cleanup > when there is no activity

<details>
<summary>Expand output</summary>

  ● Graceful Shutdown › production (standalone mode) › should not accept new requests during shutdown cleanup › when there is no activity

    expect(received).toBe(expected) // Object.is equality

    Expected: 0
    Received: null

      228 |         // App finally shuts down
      229 |         await appKilledPromise
    > 230 |         expect(app.exitCode).toBe(0)
          |                              ^
      231 |       })
      232 |     })
      233 |   }

      at Object.toBe (production/graceful-shutdown/index.test.ts:230:30)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

samcx commented:
@PatNeedham Looks like that test fail (`pnpm test-start test/production/graceful-shutdown/index.test.ts`) was a flaky fail.

PatNeedham commented:
@samcx any suggestion for how @Leonils be added as a co-author on this PR so he gets credit for resolving the initial bug?

samcx commented:
@PatNeedham Unfortunately GitHub doesn't provide a way for this besides adding a co-author to a commit. But you can add it to the <picture data-single-emoji=":pr:" title=":pr:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/pr/876af7e7d65e1595.png" alt=":pr:" align="absmiddle"></picture> description!

samcx commented:
@PatNeedham I think we're missing the test from the previous <picture data-single-emoji=":pr:" title=":pr:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/pr/876af7e7d65e1595.png" alt=":pr:" align="absmiddle"></picture>

PatNeedham commented:
Good catch, let me add that back and update this branch shortly. In the mean time, what would be the right way to run just a subset of the e2e tests locally? For example, from the [original PR's test checks](https://github.com/vercel/next.js/pull/59282/checks), only these ones failed:

- [test dev (2/4) / build](https://github.com/vercel/next.js/actions/runs/7179317638/job/19549288692?pr=59282)
- [test ppr dev (1/4) / build](https://github.com/vercel/next.js/actions/runs/7179317638/job/19549295322?pr=59282)
- [test ppr prod (2/4) / build](https://github.com/vercel/next.js/actions/runs/7179317638/job/19549297447?pr=59282)

I see the test dev (1 through 4) processes are defined here: https://github.com/vercel/next.js/blob/v14.3.0-canary.30/.github/workflows/build_and_test.yml#L287-L300 - so I tried running this command locally:

```
NEXT_TEST_MODE=dev node run-tests.js --timings -g 2/4 -c 8 --type development
```
(replacing TEST_CONCURRENCY with 8 since it's specified in the `env` section further up in the file)

That's running now for the past ~15 minutes still ongoing. I'm guessing there's a `-- --matchingTests opentelemetry` type of argument I need to include to limit which ones get run? 

I can't fully add back the `it('should handle client-side navigation', async () => { ... })` test because the supporting helper functions use `traceFile` from the ./constants file, but that value got removed a couple months back: https://github.com/vercel/next.js/pull/62120/files#diff-95fdfa81f6699cf4ad4345cd01cc66ae151d91091ddf3f5ea4f04dd714c49f1b - is it safe to re-add it to that file? 

```
export const traceFile = 'otel-trace.txt'
```

Leonils commented:
Back then I used `pnpm testheadless test/e2e/opentelemetry/opentelemetry-client-navigation.test.ts` to develop the test, but when I tried to run more tests, I took also ages...

For the test "should handle client-side navigation", it's essentially a snapshot test. I ran the test first expecting nothing, reviewed the traces and was happy with what I got. I saw some changes in theses files last time, so there might be new/modified traces too. The only important part is that we have `target` with the actual value (`"http.target": "/app/foo/loading/page1"`) and `route` with the placeholder: `"next.route": "/app/[param]/loading/page1"`

samcx commented:
@PatNeedham I can confirm the previously failed run were test flakes so that's fine. No need to merge canary as well—I'll merge it if necessary.

Leonils commented:
I doubt reintroducing the file will be enough : the TestReporter that was reporting the the file has been rewritten to send to a mock server. 

See what was done in the instrumentation when I wrote my PR versus now : 

- *Old (when I wrote my PR)*: https://github.com/Leonils/next.js/blob/697f99e1fd9d9e8bd3728e0872deb5dfcadeff5d/test/e2e/opentelemetry/instrumentation-node-test.ts
- *New*: https://github.com/vercel/next.js/blob/canary/test/e2e/opentelemetry/instrumentation-test.ts

PatNeedham commented:
Last commit I just pushed is as close as I've been to get everything passing. 
```
pnpm testheadless test/e2e/opentelemetry/opentelemetry.test.ts
```
Running that locally shows `2 failed, 25 passed, 27 total`

The 2 failed ones are just
```
opentelemetry › incoming context propagation › app router › should handle client-side navigation
opentelemetry › root context › app router › should handle client-side navigation
```

And it looks like the issue is just contained to the parentTraceId value:

![image](https://github.com/vercel/next.js/assets/967811/c1b458d5-b890-46e3-8fc5-cca77af90c6c)

What might be going awry in that test? @dvoytenko I want to loop you in on this since I see you introduced that `expectTrace` function a couple months back as part of https://github.com/vercel/next.js/pull/62120

PatNeedham commented:
For the `root context › app router` path invocation of that test, it looks like the value received as `"parentId": undefined,` in all places while the expected value was `"parentId": "<random-id>",`.

dvoytenko commented:
> Last commit I just pushed is as close as I've been to get everything passing.
> 
> ```
> pnpm testheadless test/e2e/opentelemetry/opentelemetry.test.ts
> ```
> 
> Running that locally shows `2 failed, 25 passed, 27 total`
> 
> The 2 failed ones are just
> 
> ```
> opentelemetry › incoming context propagation › app router › should handle client-side navigation
> opentelemetry › root context › app router › should handle client-side navigation
> ```
> 
> And it looks like the issue is just contained to the parentTraceId value:
> 
> ![image](https://private-user-images.githubusercontent.com/967811/327310382-c1b458d5-b890-46e3-8fc5-cca77af90c6c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU3Mjc3MDEsIm5iZiI6MTcxNTcyNzQwMSwicGF0aCI6Ii85Njc4MTEvMzI3MzEwMzgyLWMxYjQ1OGQ1LWI4OTAtNDZlMy04ZmM1LWNjYTc3YWY5MGM2Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxNFQyMjU2NDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMTFlOTAxOTJkZDc2NzUzNzc3YTU0ZmEwZmZkYWMwMjRkMWE5ZWZhNzdmNjhkZWJmZDc5NzRlOWNhMjlkOGY0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.xUDk1anLwD_sCd19XCmyfKzOh8dmupBBJV_i--FgfYo)
> 
> What might be going awry in that test? @dvoytenko I want to loop you in on this since I see you introduced that `expectTrace` function a couple months back as part of #62120

IMHO those tests just need to change to use snapshots.

PatNeedham commented:
Thank you @dvoytenko for the suggestion, I've modified the test to use that approach and updated this PR's description so I believe it's in a mergeable state now.

dvoytenko commented on packages/next/src/server/app-render/app-render.tsx:
```diff
@@ -877,6 +877,8 @@ async function renderToHTMLOrFlightImpl(
     res,
   }
 
+  getTracer().getRootSpanAttributes()?.set('next.route', pagePath)
```
Why move it up? Are some steps below sensitive to it?

Overall, the `pagePath` is not a great value here and it takes some time to update it to the right value. IMHO, too late.

PatNeedham replied:
> Why move it up? Are some steps below sensitive to it?

Yes - the `if (isRSCRequest && !isStaticGeneration) { ... }` check is where this function can end early, which affects whether `next.route` gets set. 

But also, the snapshot testing looks problematic, for example this is from the most recent PR build-and-test logs:

![image](https://github.com/vercel/next.js/assets/967811/b62d27f0-72ac-4cbc-b5d6-9f6b226c4267)

id, parentId, timestamp, traceId & duration values appear dependent on something that's preventing consistent results. Maybe [jest.setSystemTime[(https://jestjs.io/docs/jest-object#jestsetsystemtimenow-number--date) could address the timestamp part? For the others I'm not sure what might ensure that consistency.

dvoytenko replied:
We can definitely just use `matchAny` for timestamp and duration. And probably for traceId. However, id and parentId are not so simple - they represent the nesting hierarchy which is somewhat important for telemetry. We can matchAny them too, but maybe we can find a way to link them up? E.g. maybe we can drop id/parentId, but instead add a synthetic parentName? That way we can test hierarchy as well.

harry-gocity commented:
Is there anything that can be done here to get this merged? Would love to see it land soon.

tredman commented:
Would also like to see this fixed. For now we're using an OTEL span processor to reduce cardinality but it's less than ideal because we lose route information

```yaml
  span/nextjs:
    include:
      attributes:
      - key: next.span_type
        value: (BaseServer\.handleRequest|AppRender\.fetch)
      match_type: regexp
    name:
      to_attributes:
        rules:
        - ^(RSC|fetch)?.?(GET|POST|PUT) (?P<url>.+)$
```

Starefossen commented:
Are there anything we can help with to get this merged @dvoytenko, @Blackskar875 ?
